File: src-tauri/src/sessions.rs
use rusqlite::params;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::AppState;

// ─── Data types ──────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Message {
    pub id: String,
    pub session_id: String,
    pub role: String,
    pub content: String,
    pub tool_calls: Vec<serde_json::Value>,
    pub created_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Session {
    pub id: String,
    pub project_id: String,
    pub title: String,
    pub status: String,
    pub created_at: String,
    pub messages: Vec<Message>,
}

fn now_iso() -> String {
    chrono::Utc::now().to_rfc3339()
}

// ─── Tauri commands ───────────────────────────────────────────────────────────

/// Create a new idle session for a project.
#[tauri::command]
pub fn create_session(
    project_id: String,
    state: tauri::State<AppState>,
) -> Result<Session, String> {
    let id = Uuid::new_v4().to_string();
    let created_at = now_iso();

    let conn = state.db.lock().map_err(|e| e.to_string())?;
    conn.execute(
        "INSERT INTO sessions (id, project_id, title, status, created_at) VALUES (?1, ?2, 'New session', 'idle', ?3)",
        params![id, project_id, created_at],
    )
    .map_err(|e| e.to_string())?;

    Ok(Session {
        id,
        project_id,
        title: "New session".into(),
        status: "idle".into(),
        created_at,
        messages: vec![],
    })
}

/// List all sessions for a project (metadata only — messages are not loaded).
#[tauri::command]
pub fn list_sessions(
    project_id: String,
    state: tauri::State<AppState>,
) -> Result<Vec<Session>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;

    let mut stmt = conn
        .prepare(
            "SELECT id, project_id, title, status, created_at
             FROM sessions
             WHERE project_id = ?1
             ORDER BY created_at ASC",
        )
        .map_err(|e| e.to_string())?;

    let sessions = stmt
        .query_map(params![project_id], |row| {
            Ok(Session {
                id: row.get(0)?,
                project_id: row.get(1)?,
                title: row.get(2)?,
                status: row.get(3)?,
                created_at: row.get(4)?,
                messages: vec![],
            })
        })
        .map_err(|e| e.to_string())?
        .filter_map(|r| r.ok())
        .collect();

    Ok(sessions)
}

/// Fetch a single session with its full message history.
#[tauri::command]
pub fn get_session(
    session_id: String,
    state: tauri::State<AppState>,
) -> Result<Session, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;

    let mut session = conn
        .query_row(
            "SELECT id, project_id, title, status, created_at FROM sessions WHERE id = ?1",
            params![session_id],
            |row| {
                Ok(Session {
                    id: row.get(0)?,
                    project_id: row.get(1)?,
                    title: row.get(2)?,
                    status: row.get(3)?,
                    created_at: row.get(4)?,
                    messages: vec![],
                })
            },
        )
        .map_err(|e| e.to_string())?;

    let mut msg_stmt = conn
        .prepare(
            "SELECT id, session_id, role, content, created_at
             FROM messages
             WHERE session_id = ?1
             ORDER BY created_at ASC",
        )
        .map_err(|e| e.to_string())?;

    session.messages = msg_stmt
        .query_map(params![session_id], |row| {
            Ok(Message {
                id: row.get(0)?,
                session_id: row.get(1)?,
                role: row.get(2)?,
                content: row.get(3)?,
                tool_calls: vec![],
                created_at: row.get(4)?,
            })
        })
        .map_err(|e| e.to_string())?
        .filter_map(|r| r.ok())
        .collect();

    Ok(session)
}

/// Hard-delete a session and its messages (cascade via foreign key).
#[tauri::command]
pub fn delete_session(
    session_id: String,
    state: tauri::State<AppState>,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    conn.execute("DELETE FROM sessions WHERE id = ?1", params![session_id])
        .map_err(|e| e.to_string())?;
    Ok(())
}

/// Persist a single message to SQLite and return it with its generated id.
#[tauri::command]
pub fn save_message(
    session_id: String,
    role: String,
    content: String,
    state: tauri::State<AppState>,
) -> Result<Message, String> {
    let id = Uuid::new_v4().to_string();
    let created_at = now_iso();

    let conn = state.db.lock().map_err(|e| e.to_string())?;
    conn.execute(
        "INSERT INTO messages (id, session_id, role, content, created_at) VALUES (?1, ?2, ?3, ?4, ?5)",
        params![id, session_id, role, content, created_at],
    )
    .map_err(|e| e.to_string())?;

    Ok(Message {
        id,
        session_id,
        role,
        content,
        tool_calls: vec![],
        created_at,
    })
}

/// Update the session title (called after the first user message is processed).
#[tauri::command]
pub fn update_session_title(
    session_id: String,
    title: String,
    state: tauri::State<AppState>,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    conn.execute(
        "UPDATE sessions SET title = ?1 WHERE id = ?2",
        params![title, session_id],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

File: electron/sessions.ts
import * as crypto from "crypto";
import type { Database } from "better-sqlite3";
import type { Message, Session, ToolCall, ToolCallStatus, ToolCategory } from "../src/types/index";

// ─── Helpers ─────────────────────────────────────────────────────────────────

function nowIso(): string {
  return new Date().toISOString();
}

// ─── Exported functions ──────────────────────────────────────────────────────

/**
 * Create a new idle session for a project.
 * Provider and model are inherited from the project config at creation time
 * so each session remembers which model it was started with.
 */
export function createSession(
  db: Database,
  projectId: string,
  provider: string | null = null,
  model: string | null = null
): Session {
  const id = crypto.randomUUID();
  const createdAt = nowIso();

  db.prepare(
    "INSERT INTO sessions (id, project_id, title, status, created_at, provider, model) VALUES (?, ?, 'New session', 'idle', ?, ?, ?)"
  ).run(id, projectId, createdAt, provider, model);

  return {
    id,
    project_id: projectId,
    title: "New session",
    status: "idle",
    created_at: createdAt,
    messages: [],
    provider,
    model,
    agent: null,
    skills: null,
  };
}

/**
 * List all sessions for a project (metadata only — messages are not loaded).
 */
export function listSessions(db: Database, projectId: string): Session[] {
  const rows = db
    .prepare(
      `SELECT id, project_id, title, status, created_at, provider, model, agent, skills
       FROM sessions
       WHERE project_id = ?
       ORDER BY created_at ASC`
    )
    .all(projectId) as {
    id: string;
    project_id: string;
    title: string;
    status: string;
    created_at: string;
    provider: string | null;
    model: string | null;
    agent: string | null;
    skills: string | null;
  }[];

  return rows.map((row) => ({
    id: row.id,
    project_id: row.project_id,
    title: row.title,
    status: row.status as Session["status"],
    created_at: row.created_at,
    messages: [],
    provider: row.provider,
    model: row.model,
    agent: row.agent,
    skills: row.skills ? (JSON.parse(row.skills) as string[]) : null,
  }));
}

/**
 * Fetch a single session with its full message history.
 */
export function getSession(db: Database, sessionId: string): Session {
  const row = db
    .prepare(
      "SELECT id, project_id, title, status, created_at, provider, model, agent, skills FROM sessions WHERE id = ?"
    )
    .get(sessionId) as
    | {
        id: string;
        project_id: string;
        title: string;
        status: string;
        created_at: string;
        provider: string | null;
        model: string | null;
        agent: string | null;
        skills: string | null;
      }
    | undefined;

  if (!row) {
    throw new Error(`Session not found: ${sessionId}`);
  }

  const messageRows = db
    .prepare(
      `SELECT id, session_id, role, content, created_at, agent
       FROM messages
       WHERE session_id = ?
       ORDER BY created_at ASC`
    )
    .all(sessionId) as {
    id: string;
    session_id: string;
    role: string;
    content: string;
    created_at: string;
    agent: string | null;
  }[];

  // Load tool_calls for all messages in one query and group by message_id
  const toolCallsByMessageId = new Map<string, ToolCall[]>();
  if (messageRows.length > 0) {
    const messageIds = messageRows.map((m) => m.id);
    const placeholders = messageIds.map(() => "?").join(", ");
    const toolCallRows = db
      .prepare(
        `SELECT id, message_id, name, args, result, status, category
         FROM tool_calls
         WHERE message_id IN (${placeholders})
         ORDER BY rowid ASC`
      )
      .all(...messageIds) as {
      id: string;
      message_id: string;
      name: string;
      args: string;
      result: string | null;
      status: string;
      category: string;
    }[];

    for (const row of toolCallRows) {
      const tc: ToolCall = {
        id: row.id,
        name: row.name,
        args: JSON.parse(row.args) as Record<string, unknown>,
        result: row.result ? (JSON.parse(row.result) as unknown) : null,
        status: row.status as ToolCallStatus,
        category: row.category as ToolCategory,
      };
      const existing = toolCallsByMessageId.get(row.message_id) ?? [];
      existing.push(tc);
      toolCallsByMessageId.set(row.message_id, existing);
    }
  }

  const messages: Message[] = messageRows.map((m) => ({
    id: m.id,
    session_id: m.session_id,
    role: m.role as Message["role"],
    content: m.content,
    tool_calls: toolCallsByMessageId.get(m.id) ?? [],
    created_at: m.created_at,
    agent: m.agent,
  }));

  return {
    id: row.id,
    project_id: row.project_id,
    title: row.title,
    status: row.status as Session["status"],
    created_at: row.created_at,
    messages,
    provider: row.provider,
    model: row.model,
    agent: row.agent,
    skills: row.skills ? (JSON.parse(row.skills) as string[]) : null,
  };
}

/**
 * Hard-delete a session and its messages (cascade via foreign key).
 */
export function deleteSession(db: Database, sessionId: string): void {
  db.prepare("DELETE FROM sessions WHERE id = ?").run(sessionId);
}

/**
 * Persist the runtime status of a session to SQLite.
 * Called by the agent runner at each status transition so restarts can
 * detect and recover sessions that were interrupted mid-turn.
 */
export function updateSessionStatus(
  db: Database,
  sessionId: string,
  status: Session["status"]
): void {
  db.prepare("UPDATE sessions SET status = ? WHERE id = ?").run(status, sessionId);
}

/**
 * On startup, mark any session still in "running" status as "error".
 * A session can only be "running" if the app crashed or was force-quit
 * mid-turn — it will never complete, so surface it as an error.
 * Returns the number of sessions recovered.
 */
export function recoverStaleSessionStatuses(db: Database): number {
  const result = db
    .prepare("UPDATE sessions SET status = 'error' WHERE status = 'running'")
    .run();
  return result.changes;
}

/**
 * Persist a single message to SQLite and return it with its generated id.
 */
export function saveMessage(
  db: Database,
  args: { sessionId: string; role: string; content: string; agent?: string | null }
): Message {
  const id = crypto.randomUUID();
  const createdAt = nowIso();

  db.prepare(
    "INSERT INTO messages (id, session_id, role, content, created_at, agent) VALUES (?, ?, ?, ?, ?, ?)"
  ).run(id, args.sessionId, args.role, args.content, createdAt, args.agent ?? null);

  return {
    id,
    session_id: args.sessionId,
    role: args.role as Message["role"],
    content: args.content,
    tool_calls: [],
    created_at: createdAt,
    agent: args.agent ?? null,
  };
}

/**
 * Update the session title (called after the first user message is processed).
 */
export function updateSessionTitle(
  db: Database,
  sessionId: string,
  title: string
): void {
  db.prepare("UPDATE sessions SET title = ? WHERE id = ?").run(title, sessionId);
}

/**
 * Update the provider and model for a session (changed from the model picker).
 */
export function updateSessionModel(
  db: Database,
  sessionId: string,
  provider: string,
  model: string
): void {
  db.prepare("UPDATE sessions SET provider = ?, model = ? WHERE id = ?").run(
    provider,
    model,
    sessionId
  );
}

/**
 * Update the selected sub-agent for a session (null = default agent).
 */
export function updateSessionAgent(
  db: Database,
  sessionId: string,
  agent: string | null
): void {
  db.prepare("UPDATE sessions SET agent = ? WHERE id = ?").run(agent, sessionId);
}

/**
 * Update the active skills for a session (empty array = no skills active).
 */
export function updateSessionSkills(
  db: Database,
  sessionId: string,
  skills: string[]
): void {
  const value = skills.length > 0 ? JSON.stringify(skills) : null;
  db.prepare("UPDATE sessions SET skills = ? WHERE id = ?").run(value, sessionId);
}

// ── Tool call persistence ─────────────────────────────────────────────────────

/**
 * Creates an empty placeholder assistant message at turn start.
 * All tool calls for this turn will reference this message via FK.
 * Returns the newly created message.
 */
export function createPlaceholderMessage(
  db: Database,
  args: { sessionId: string; agent?: string | null }
): Message {
  const id = crypto.randomUUID();
  const createdAt = nowIso();

  db.prepare(
    "INSERT INTO messages (id, session_id, role, content, created_at, agent) VALUES (?, ?, ?, ?, ?, ?)"
  ).run(id, args.sessionId, "assistant", "", createdAt, args.agent ?? null);

  return {
    id,
    session_id: args.sessionId,
    role: "assistant",
    content: "",
    tool_calls: [],
    created_at: createdAt,
    agent: args.agent ?? null,
  };
}

/**
 * Updates an existing message's content (called at turn end with full text).
 */
export function updateMessageContent(
  db: Database,
  messageId: string,
  content: string
): void {
  db.prepare("UPDATE messages SET content = ? WHERE id = ?").run(content, messageId);
}

/**
 * Inserts a tool_calls row. Call when a tool starts executing.
 */
export function saveToolCall(
  db: Database,
  args: {
    id: string;
    messageId: string;
    name: string;
    args: Record<string, unknown>;
    status: ToolCallStatus;
    category: string;
  }
): void {
  db.prepare(
    "INSERT INTO tool_calls (id, message_id, name, args, result, status, category) VALUES (?, ?, ?, ?, NULL, ?, ?)"
  ).run(
    args.id,
    args.messageId,
    args.name,
    JSON.stringify(args.args),
    args.status,
    args.category
  );
}

/**
 * Updates status (and optionally result) of an existing tool_call row.
 */
export function updateToolCallStatus(
  db: Database,
  id: string,
  status: ToolCallStatus,
  result?: unknown
): void {
  if (result !== undefined) {
    db.prepare("UPDATE tool_calls SET status = ?, result = ? WHERE id = ?").run(
      status,
      JSON.stringify(result),
      id
    );
  } else {
    db.prepare("UPDATE tool_calls SET status = ? WHERE id = ?").run(status, id);
  }
}

/**
 * Loads all tool_calls for a message, ordered by insertion.
 */
export function loadToolCallsForMessage(db: Database, messageId: string): ToolCall[] {
  const rows = db
    .prepare(
      "SELECT id, name, args, result, status, category FROM tool_calls WHERE message_id = ? ORDER BY rowid ASC"
    )
    .all(messageId) as {
    id: string;
    name: string;
    args: string;
    result: string | null;
    status: string;
    category: string;
  }[];

  return rows.map((row) => ({
    id: row.id,
    name: row.name,
    args: JSON.parse(row.args) as Record<string, unknown>,
    result: row.result ? (JSON.parse(row.result) as unknown) : null,
    status: row.status as ToolCallStatus,
    category: row.category as ToolCategory,
  }));
}
